pkg/openstack/loadbalancer.go

```diff
@@ -2078,14 +2101,21 @@ func (lbaas *LbaasV2) ensureSecurityRule(
 	}
 
 	mc := metrics.NewMetricContext("security_group_rule", "create")
-	_, err := rules.Create(lbaas.network, sgRuleCreateOpts).Extract()
+	sg, err := rules.Create(lbaas.network, sgRuleCreateOpts).Extract()
 	if err != nil && cpoerrors.IsConflictError(err) {
 		// Conflict means the SG rule already exists, so ignoring that error.
-		return mc.ObserveRequest(nil)
+		// We can get the rule ID from the message.
+		re := regexp.MustCompile("Rule id is ([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})")
```

Comment:
Please put MustCompile() in a static context. If it's going to panic it should do it when the module is loaded. Also, parsing human readable error messages is nasty.

This looks like an effort to do atomic creation, which I'm normally very much in favour of, but I'm not sure it's serving us here. I think this code would be a lot cleaner if we fetched all rules first and did creation/deletion based on that. The controller should already prevent the race of simultaneous reconciliation of a single object, but even if it didn't we should be safe if we always fail on a 409 here.